### ISSUE TITLE
Allow ignoring seasonal roads

### ISSUE THREAD

**Issue by PureTryOut** [graphhopper/graphhopper#2477]
Right now I'm planning for a roadtrip next summer, Graphhopper so far has been by far the best tool for this that I found. This route will take us through Norway which has lots of roads only open in the summer. Since I'm planning for the summer, this isn't a problem but I'm making this route in a season in which those roads are closed, and thus Graphhopper refuses to calculate routes on those roads. It makes it hard to plan months in advance as the route as I make it now will look different when we are actually driving it.

I'd like an option to ignore seasonal restrictions and just route over such a road, no matter what season I'm in.

**Comment by karussell** [comment]
Yeah, we should probably lift those restrictions and put this in a special encoded value so that one can still exclude this. Related to #1260.

Do you have a specific example?

**Comment by PureTryOut** [comment]
Yes, for example https://graphhopper.com/maps/?point=62.082391%2C7.227845&point=61.939839%2C7.261105&point=62.015004%2C7.647686&locale=en-us&elevation=true&profile=car&use_miles=false&selected_detail=Elevation&layer=Omniscale

Instead of following the road straight down south from the start, it goes up north instead and drives all the way around. If you however do the same route in the summer, it follows the route down south just fine.

**Comment by otbutz** [comment]
The biggest question is the detail level we can or want to provide if we add an encoded value for this.

Conditional access rules can be arbitrarily complex. For your route it's a rather tame example: `access:conditional=no @ (Nov-May)` but things can get quite messy: `motor_vehicle:conditional=delivery @ (Mo-Fr 06:00-11:00,17:00-19:00;Sa 03:30-19:00)`

For time based rules we'd also need to resolve the applicable timezone...

**Comment by michaz** [comment]
We [researched this extensively](https://www.bmvi.de/SharedDocs/DE/Artikel/DG/mfund-projekte/tardur.html). Some of us might remember.

Parsing those rules is messy, but has been done to some level of detail. We could dig out the code. Question is what happens then.

Do we evaluate the rules at import time, and say, when you import in summer, you get a summer graph? That's easy.

Do we evaluate them at query time, so the user specifies a calendar date for which the route is going to be planned? That doesn't work with pre-preparation of any kind -- you would need one profile per day basically, as restrictions overlap.

With finer grained restrictions (time of day) it's even worse, as "real" time-dependent routing (where the clock moves forward as you go) is a different thing altogether.

My conclusion back then was that this is something for hand-built, curated scenarios (for a region), and we could (in principle, given resources) provide the tools for this. But supporting this as a world-wide feature in full generality is not practical at the moment.

**Comment by michaz** [comment]
*Or* we could go and update our findings to reflect that we now have CCHs, and see how far that gets us. :-)

**Comment by PureTryOut** [comment]
> I think OP just had a switch in mind where you can say "ignore seasonal restrictions".

Yes I did :wink: Time of day access is of course way more complex and not something I'm personally interested in. I was just thinking of being able to set a date the route is being planned in, and then it'll use the seasonal restrictions of that date. Of course if a route takes e.g. 3 days to complete, that 3rd day might be in another season causing new road restrictions, but that is again more complex than I personally need.

**Comment by otbutz** [comment]
> I think OP just had a switch in mind where you can say "ignore seasonal restrictions". That's easy of course.

We would need a 12bit EV to encode the months as a bitset. Seasons are a regional thing, so we can't use something like `Season.WINTER`.

**Comment by karussell** [comment]
We could use a value with a single bit that just says this is a "currently active recurring restriction":

1. This would fix this issue without a lot of additional data as it just lifts the restriction
2. It would be still possible to exclude current recurring (seasonal) restrictions per request
3. We would not lift the restriction in case there is a onetime conditional restriction like a road closure

**Comment by otbutz** [comment]
I am not very happy with the concept that the determined value is dependent on the import time. This might be perfectly fine for the GH project as the graph is updated frequently. For users who like to plan their routes in advance like OP or others who reimport their graph not that frequently, the routing might be off.

My proposal: just mark them as seasonal true/false if their conditional access is based on months. Ideally we should use this information for the instructions and to allow them to be used in custom profiles.

**Comment by hungerburg** [comment]
Will this also work here - https://graphhopper.com/maps/?point=47.169297%2C11.464763&point=47.176746%2C11.503031&profile=bike&layer=TF+Cycle

The seasonal MTB route is tagged with bicycle:conditional=yes @(Apr-Oct) - but there is also vehicle=no specified -- so to really be useful in the sense of this issue, the conditional would have to be converted to unconditional upfront?

**Comment by hungerburg** [comment]
There is also a gate https://www.openstreetmap.org/node/1640506593 there - might that cause the detour?
